geomstats/geometry/trees.py

```diff
@@ -0,0 +1,473 @@
+            b1 = set(u).issubset(set(self.part1)) and set(v).issubset(set(self.part2))
+            b2 = set(v).issubset(set(self.part1)) and set(u).issubset(set(self.part2))
+            return b1 or b2
+
+    def point_to_split(self, other):
+        """ Gives back the part of this split that is directed to the other split. """
+        return self.part1 if set(self.part1) & set(other.part1) and set(
+            self.part1) & set(other.part2) else self.part2
+
+    def point_away_split(self, other):
+        """ Gives back the part of this split that is directed away from other split."""
+        return self.part2 if set(self.part1) & set(other.part1) and set(
+            self.part1) & set(other.part2) else self.part1
+
+    def compatible_with(self, other):
+        """ Checks whether this split is compatible with another split. """
+        p1, p2 = set(self.part1), set(self.part2)
+        o1, o2 = set(other.part1), set(other.part2)
+        return not gs.all([p1 & o1, p1 & o2, p2 & o1, p2 & o2])
+
+    def __eq__(self, other):
+        """ The equal function. """
+        return self.__hash__() == other.__hash__()
+
+    def __lt__(self, other):
+        """ Less than function (<). """
+        return self.__hash__() < other.__hash__()
+
+    def __hash__(self):
+        """ The hash function. """
+        return hash((self.n, self._part1, self._part2))
+
+    def __str__(self):
+        """ String representation of the split. """
+        return str((self.part1, self.part2))
+
+    def __repr__(self):
+        return f"{self.part1}|{self.part2}"
+
+    def __bool__(self):
+        """ Returns False only if both parts are empty sets. """
+        return bool(self.part1) and bool(self.part2)
+
+
+class Structure(object):
+    r""" A structure of a forest, that is the connected components and the edges.
+
+    Parameters
+    ----------
+    n : int
+        Number of labels, the set of labels is then :math:`\{0,\dots,n-1\}`.
+    partition :
+        An iterable of iterables, namely a partition of the set :math:`\{0,\dots,n-1\}`,
+        representing the connected components of the forest.
+    split_sets :
+        An iterable of iterables of splits, where each set of splits contains only
+        splits of the respective connected component in the partition, so their order
+        is related. The splits are then the edges of the connected components,
+        respectively, and thus the union of all sets of splits is all edges of the
+        forest.
+    """
+
+    def __init__(self, n, partition, split_sets):
+        """ Initialize a structure with partition and sets of splits of components. """
+        self._n = n
+        # make some assertions about the given parameters.
+        # same number of components in both partition and split_sets
+        assert len(split_sets) == len(partition)
+        # all parts of partition give the whole set of labels.
+        assert set.union(*[set(part) for part in partition]) == set(range(n))
```

Comment:
We usually raise value errors instead of using assert statement.

Reply:
I've replaced the assert statements with if-clauses and ValueErrors